pkg/runtime/wfengine/activity.go

```diff
@@ -50,15 +72,68 @@ func (a *activityActor) SetActorRuntime(actorsRuntime actors.Actors) {
-	err := a.createReliableReminder(ctx, actorID, methodName, data)
+	var ar ActivityRequest
+	if err := actors.DecodeInternalActorData(data, &ar); err != nil {
+		return nil, fmt.Errorf("failed to decode activity request: %w", err)
+	}
+
+	// Try to load activity state. If we find any, that means the activity invocation is a duplicate.
+	if state, err := a.loadActivityState(ctx, actorID, ar.Generation); err != nil {
+		return nil, err
+	} else if state.Generation > 0 {
+		return nil, ErrDuplicateInvocation
+	}
+
+	// Save the request details to the state store in case we need it after recovering from a failure.
+	state := activityState{
+		Generation:   ar.Generation,
+		EventPayload: ar.HistoryEvent,
+	}
+	if err := a.saveActivityState(ctx, actorID, state); err != nil {
+		return nil, err
+	}
+
+	// The actual execution is triggered by a reminder
+	err := a.createReliableReminder(ctx, actorID, "run-activity", ar.Generation)
 	return nil, err
 }
 
 // InvokeReminder implements actors.InternalActor and executes the activity logic.
-// TODO: Need to review how errors for reminders are handled.
 func (a *activityActor) InvokeReminder(ctx context.Context, actorID string, reminderName string, data any, dueTime string, period string) error {
-	taskEvent, err := backend.UnmarshalHistoryEvent(data.([]byte))
+	wfLogger.Debugf("invoking reminder '%s' on activity actor '%s'", reminderName, actorID)
+
+	generation := data.(uint64)
+	state, _ := a.loadActivityState(ctx, actorID, generation)
+	// TODO: On error, reply with a failure - this requires support from durabletask-go to produce TaskFailure results
+
+	timeoutCtx, cancelTimeout := context.WithTimeout(ctx, a.defaultTimeout)
+	defer cancelTimeout()
+
+	if err := a.executeActivity(timeoutCtx, actorID, reminderName, state.EventPayload); err != nil {
+		if err == timeoutCtx.Err() {
+			wfLogger.Warnf("%s: execution of '%s' timed-out and will be retried later", actorID, reminderName)
+
+			// Returning nil signals that we want the execution to be retried in the next period interval
+			return nil
+		} else if _, ok := err.(recoverableError); ok {
+			wfLogger.Errorf("%s: execution failed with a recoverable error and will be retried later: %v", actorID, err)
+
+			// Returning nil signals that we want the execution to be retried in the next period interval
+			return nil
+		} else {
+			wfLogger.Errorf("%s: execution failed with a non-recoverable error: %v", actorID, err)
+			// TODO: Reply with a failure - this requires support from durabletask-go to produce TaskFailure results
+		}
+	}
+
+	// TODO: Purge actor state based on some data retention policy
+
+	// We delete the reminder on success and on non-recoverable errors.
+	return actors.ErrReminderCanceled
+}
+
+func (a *activityActor) executeActivity(ctx context.Context, actorID string, name string, eventPayload []byte) error {
+	taskEvent, err := backend.UnmarshalHistoryEvent(eventPayload)
 	if err != nil {
 		return err
```

Comment:
Do we want the fmt.Errorf here as well?

Reply:
I'm intentionally not doing `fmt.Errorf` here because I think the error message generated by the `UnmarshalHistoryEvent` is sufficiently descriptive. It's also an error message that we theoretically should never run into, so I don't think it's important to add additional context.